test/unit/datetime.c

```diff
@@ -0,0 +1,261 @@
+	S("2012-12-24t15:30z"),
+	S("2012-12-24t15:30"),
+	S("2012-12-24t16:30+01:00"),
+	S("2012-12-24t16:30+0100"),
+	S("2012-12-24t14:30-01:00"),
+	S("2012-12-24t14:30-0100"),
+	S("2012-12-24t15:30:00Z"),
+	S("2012-12-24t15:30:00z"),
+	S("2012-12-24t15:30:00"),
+	S("2012-12-24t16:30:00+01:00"),
+	S("2012-12-24t16:30:00+0100"),
+	S("2012-12-24t14:30:00-01:00"),
+	S("2012-12-24t14:30:00-0100"),
+	S("2012-12-24t15:30:00.123456Z"),
+	S("2012-12-24t15:30:00.123456z"),
+	S("2012-12-24t16:30:00.123456+01:00"),
+	S("2012-12-24t14:30:00.123456-01:00"),
+	S("2012-12-24 16:30 +01:00"),
+	S("2012-12-24 14:30 -01:00"),
+	S("2012-12-24 15:30 UTC"),
+	S("2012-12-24 16:30 UTC+1"),
+	S("2012-12-24 16:30 UTC+01"),
+	S("2012-12-24 16:30 UTC+0100"),
+	S("2012-12-24 16:30 UTC+01:00"),
+	S("2012-12-24 14:30 UTC-1"),
+	S("2012-12-24 14:30 UTC-01"),
+	S("2012-12-24 14:30 UTC-01:00"),
+	S("2012-12-24 14:30 UTC-0100"),
+	S("2012-12-24 15:30 GMT"),
+	S("2012-12-24 16:30 GMT+1"),
+	S("2012-12-24 16:30 GMT+01"),
+	S("2012-12-24 16:30 GMT+0100"),
+	S("2012-12-24 16:30 GMT+01:00"),
+	S("2012-12-24 14:30 GMT-1"),
+	S("2012-12-24 14:30 GMT-01"),
+	S("2012-12-24 14:30 GMT-01:00"),
+	S("2012-12-24 14:30 GMT-0100"),
+	S("2012-12-24 14:30 -01:00"),
+	S("2012-12-24 16:30:00 +01:00"),
+	S("2012-12-24 14:30:00 -01:00"),
+	S("2012-12-24 16:30:00.123456 +01:00"),
+	S("2012-12-24 14:30:00.123456 -01:00"),
+	S("2012-12-24 15:30:00.123456 -00:00"),
+	S("20121224T1630+01:00"),
+	S("2012-12-24T1630+01:00"),
+	S("20121224T16:30+01"),
+	S("20121224T16:30 +01"),
+};
+#undef S
+
+static int
+parse_datetime(const char *str, size_t len, int64_t *secs_p,
+	       int32_t *nanosecs_p, int32_t *offset_p)
+{
+	size_t n;
+	dt_t dt;
+	char c;
+	int sec_of_day = 0, nanosecond = 0, offset = 0;
+
+	n = dt_parse_iso_date(str, len, &dt);
+	if (!n)
+		return 1;
+	if (n == len)
+		goto exit;
+
+	c = str[n++];
+	if (!(c == 'T' || c == 't' || c == ' '))
+		return 1;
+
+	str += n;
+	len -= n;
+
+	n = dt_parse_iso_time(str, len, &sec_of_day, &nanosecond);
+	if (!n)
+		return 1;
+	if (n == len)
+		goto exit;
+
+	if (str[n] == ' ')
+		n++;
+
+	str += n;
+	len -= n;
+
+	n = dt_parse_iso_zone_lenient(str, len, &offset);
+	if (!n || n != len)
+		return 1;
+
+exit:
+	*secs_p = ((int64_t)dt_rdn(dt) - DT_EPOCH_1970_OFFSET) * SECS_PER_DAY +
+		  sec_of_day - offset * 60;
+	*nanosecs_p = nanosecond;
+	*offset_p = offset;
+
+	return 0;
+}
+
+static int
+local_rd(const struct datetime *dt)
+{
+	return (int)((int64_t)dt->epoch / SECS_PER_DAY) + DT_EPOCH_1970_OFFSET;
+}
+
+static int
+local_dt(const struct datetime *dt)
+{
+	return dt_from_rdn(local_rd(dt));
+}
+
+struct tm *
+datetime_to_tm(struct datetime *dt)
+{
+	static struct tm tm;
+
+	memset(&tm, 0, sizeof(tm));
+	dt_to_struct_tm(local_dt(dt), &tm);
+
+	int seconds_of_day = (int64_t)dt->epoch % 86400;
+	tm.tm_hour = (seconds_of_day / 3600) % 24;
+	tm.tm_min = (seconds_of_day / 60) % 60;
+	tm.tm_sec = seconds_of_day % 60;
+
+	return &tm;
+}
+
+static void datetime_test(void)
```

Comment:
We normally put return type on a separate line. Please, conform.

Reply:
Fixed. [These kinds of complains are usually an indication that we need to run clang-format over the code, so here are results]...

```diff
commit 0c6664d82a6055d364f620491206a5d4145bbe38
Author: Timur Safin <tsafin@tarantool.org>
Date:   Sun Sep 26 20:13:33 2021 +0300

    temp - clangformat for datetime_test function from test/unit/datetime.c

diff --git a/test/unit/datetime.c b/test/unit/datetime.c
index 73e276c9a..d3b1d4222 100644
--- a/test/unit/datetime.c
+++ b/test/unit/datetime.c
@@ -4,9 +4,9 @@
 #include <string.h>
 #include <time.h>
 
-#include "unit.h"
 #include "datetime.h"
 #include "trivia/util.h"
+#include "unit.h"
 
 static const char sample[] = "2012-12-24T15:30Z";
 
@@ -164,7 +164,8 @@ datetime_to_tm(struct datetime *dt)
 	return &tm;
 }
 
-static void datetime_test(void)
+static void
+datetime_test(void)
 {
 	size_t index;
 	int64_t secs_expected;
@@ -172,8 +173,8 @@ static void datetime_test(void)
 	int32_t offset;
 
 	plan(355);
-	parse_datetime(sample, sizeof(sample) - 1,
-		       &secs_expected, &nanosecs, &offset);
+	parse_datetime(sample, sizeof(sample) - 1, &secs_expected, &nanosecs,
+		       &offset);
 
 	for (index = 0; index < lengthof(tests); index++) {
 		int64_t secs;
@@ -181,8 +182,9 @@ static void datetime_test(void)
 					&secs, &nanosecs, &offset);
 		is(rc, 0, "correct parse_datetime return value for '%s'",
 		   tests[index].str);
-		is(secs, secs_expected, "correct parse_datetime output "
-					"seconds for '%s",
+		is(secs, secs_expected,
+		   "correct parse_datetime output "
+		   "seconds for '%s",
 		   tests[index].str);
 
 		/*
@@ -190,22 +192,22 @@ static void datetime_test(void)
 		 * time fields
 		 */
 		static char buff[40];
-		struct datetime dt = {secs, nanosecs, offset, 0};
+		struct datetime dt = { secs, nanosecs, offset, 0 };
 		/* datetime_to_tm returns time in GMT zone */
 		struct tm *p_tm = datetime_to_tm(&dt);
 		size_t len = strftime(buff, sizeof(buff), "%F %T", p_tm);
 		ok(len > 0, "strftime");
 		int64_t parsed_secs;
 		int32_t parsed_nsecs, parsed_ofs;
-		rc = parse_datetime(buff, len, &parsed_secs, &parsed_nsecs, &parsed_ofs);
+		rc = parse_datetime(buff, len, &parsed_secs, &parsed_nsecs,
+				    &parsed_ofs);
 		is(rc, 0, "correct parse_datetime return value for '%s'", buff);
-		is(secs, parsed_secs,
-		   "reversible seconds via strftime for '%s", buff);
+		is(secs, parsed_secs, "reversible seconds via strftime for '%s",
+		   buff);
 	}
 	check_plan();
 }
 
-
 static void
 tostring_datetime_test(void)
 {
```